### PR TITLE
Improve: add data extraction management command

### PR DESCRIPTION
**Pull request by asamolion**

### CODE REVIEW

File: openedx/features/database_extract/migrations/0004_auto_20180425_0547.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        ('database_extract', '0003_targetcourse_email_list'),
+    ]
+
+    operations = [
+        migrations.RenameField(
+            model_name='targetcourse',
+            old_name='email_list',
+            new_name='emails',
+        ),
+    ]
```
Comment (muhammadnadeem) [inline openedx/features/database_extract/migrations/0004_auto_20180425_0547.py]: delete all existing migrations and create only one migration as all db related things are final now.

File: openedx/features/database_extract/models.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    """
+    These courses are marked for data extraction. These will be used to determine data of which courses is to be extracted
+
+    @course_id: (string) id of the course
+    """
+    course_id = models.CharField(max_length=255)
+    emails = models.TextField()
+
+    def __unicode__(self):
+        return '{}'.format(self.course_id)
```
Comment (muhammadnadeem) [inline openedx/features/database_extract/models.py]: enter new line at end of each file.

File: lms/envs/common.py [hunk 3]
```diff
@@ -2229,6 +2229,9 @@
 
     # student_dashboard App
     'lms.djangoapps.student_dashboard',
+
+    # Data extraction App
+    'openedx.features.database_extract',
```
[the remark this answers is not in the document]
Comment (asamolion) [reply lms/envs/common.py]: done.

File: openedx/features/database_extract/management/commands/get_course_analytics.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+from openedx.core.djangoapps.content.course_structures.models import CourseStructure
+from openedx.features.database_extract.models import TargetCourse
+from submissions.models import StudentItem, Submission, Score
+from submissions.api import _get_or_create_student_item
+
+
+def get_file_url(answer):
+    try:
+        return ora_file_upload_api.get_download_url(answer.file_key)
+    except:
```
[the remark this answers is not in the document]
Comment (asamolion) [reply openedx/features/database_extract/management/commands/get_course_analytics.py]: used AttributeError

File: openedx/features/database_extract/management/commands/get_course_analytics.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+            course_structure = CourseStructure.objects.get(course_id=course_key)
+            course_teams = CourseTeam.objects.filter(course_id=course_key)
+
+            course_data['course_structure'] = {
+                'created': course_structure.created.__str__(),
+                'modified': course_structure.modified.__str__(),
+                'course_id': course_structure.course_id.to_deprecated_string(),
+                'structure_json': course_structure.structure_json,
+                'discussion_id_map_json': course_structure.discussion_id_map_json,
```
[the remark this answers is not in the document]
Comment (asamolion) [reply openedx/features/database_extract/management/commands/get_course_analytics.py]: No. this is something that is built in. I'm not sure what this is. Probably the discussion_id of default Ruby/Sinatra app that edx uses.

File: openedx/features/database_extract/management/commands/get_course_analytics.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+            }
+
+            for team in course_teams:
+                team_data = {
+                    'team_id': course_team.team_id,
+                    'discussion_topic_id': course_team.discussion_topic_id,
```
[the remark this answers is not in the document]
Comment (asamolion) [reply openedx/features/database_extract/management/commands/get_course_analytics.py]: the discussion_topic_id of our NodeBB discussion. Currently, we are not delivering this. Will deliver if needed.

File: openedx/features/database_extract/management/commands/get_course_analytics.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+                    'description': course_team.description,
+                    'country': course_team.country,
+                    'language': course_team.language,
+                    'last_activity_at': course_team.last_activity_at.__str__(),
+                    'team_size': course_team.team_size,
+                }
+                course_data['team_data'].push(team_data)
+
+            user_profiles = get_enrolled_students(target_course.course_id)
+            for profile in user_profiles:
```
[the remark this answers is not in the document]
Comment (asamolion) [reply openedx/features/database_extract/management/commands/get_course_analytics.py]: separated into different helper functions in helpers.py

File: openedx/features/database_extract/management/commands/get_course_analytics.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+                # get the user profie data from NodeBB API
+                data_endpoint = settings.NODEBB_ENDPOINT + '/api/v2/users/data'
+                headers = {'Authorization': 'Bearer ' + settings.NODEBB_MASTER_TOKEN}
+                response = requests.post(data_endpoint,
+                                         data={'_uid': 1, 'username': profile.user.username},
+                                         headers=headers)
+
+                user_community_data = json.loads(response._content)['payload']
+
+                demo_data = {
```
[the remark this answers is not in the document]
Comment (asamolion) [reply openedx/features/database_extract/management/commands/get_course_analytics.py]: changed to demographic_data

File: openedx/features/database_extract/management/commands/get_course_analytics.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+                    'country': profile.country.__str__(),
+                    'language': profile.language,
+                    'english_proficiency': profile.user.extended_profile.english_proficiency,
+                    'label': profile.user.extended_profile.organization.label if
+                    profile.user.extended_profile.organization else '',
+                    'reputation': user_community_data['reputation'],
+                    'postcount': user_community_data['postcount'],
+                }
+
+                anon_user_id = AnonymousUserId.objects.get(
```
[the remark this answers is not in the document]
Comment (asamolion) [reply openedx/features/database_extract/management/commands/get_course_analytics.py]: done

File: openedx/features/database_extract/management/commands/get_course_analytics.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+                        'module_type': module.module_type,
+                        'module_id': module.module_state_key.to_deprecated_string(),
+                        'course_id': module.course_id.to_deprecated_string(),
+                        'state': module.state,
+                        'grade': module.grade,
+                        'max_grade': module.max_grade,
+                        'done': module.done,
+                        'created': module.created.__str__(),
+                        'modified': module.modified.__str__(),
+                    } for module in StudentModule.objects.filter(student_id=profile.user.id)],
```
[the remark this answers is not in the document]
Comment (asamolion) [reply openedx/features/database_extract/management/commands/get_course_analytics.py]: I think they don't mind if we deliver this data as well.

File: openedx/features/database_extract/migrations/0004_auto_20180425_0547.py (hunk 1 above)
Comment (asamolion) [reply openedx/features/database_extract/migrations/0004_auto_20180425_0547.py]: done.

File: openedx/features/database_extract/models.py (hunk 2 above)
Comment (asamolion) [reply openedx/features/database_extract/models.py]: done

**Pull request merged**